### PR TITLE
Chunking input before writing a ParquetCachedBatch

### PR DESCRIPTION
**Pull request by razajafri**
This PR deals with breaking up the incoming batches before writing them to a ParquetCachedBatch to make sure the CachedBuffer never exceeds the Int max-value. 

I am still working on writing the tests to test the CPU writing. This PR so far only has tests for the `compressColumnarBatchWithParquet`

### CODE REVIEW

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 1]
```diff
@@ -253,6 +253,16 @@ private case class CloseableColumnBatchIterator(iter: Iterator[ColumnarBatch]) e
  * This class assumes, the data is Columnar and the plugin is on
  */
 class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
+
```
Comment (razajafri) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:256]: This change was made to gain access to the Inner class, open to suggestions to make this better

**Comment by razajafri** [comment]
This is pretty much done. It's still a WIP because the thing remaining is that the tests won't compile with Spark 3.0.0. I have used reflection for some tests that were simple but for others, especially the ones on the CPU, the tests looked really hard to read after I tried reflection.

@tgravescs following up on our conversation, I think we should think about alternatives like you said, maybe these tests can live with the shim310?

**Comment by razajafri** [comment]
As decided in the sprint-planning I am removing the WIP tag and marking this for review. We will include tests for the CPU part when we have a framework to compile tests selectively

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 2]
```diff
@@ -253,6 +253,16 @@ private case class CloseableColumnBatchIterator(iter: Iterator[ColumnarBatch]) e
...
+  private[rapids] def getProducer(
```
Comment (jlowe) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:257]: Unused.

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 3]
```diff
@@ -330,19 +340,58 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
...
-    withResource(GpuColumnVector.from(gpuCB)) { table =>
-      withResource(Table.writeParquetChunked(ParquetWriterOptions.DEFAULT, buffer)) { writer =>
-        writer.write(table)
+  val _2GB: Long = 2L * 1024 * 1024 * 1024
+  val APPROX_PAR_META_DATA: Int = 10 * 1024 * 1024 // we are estimating 10MB
+  val BYTES_ALLOWED_PER_BATCH: Long = _2GB - APPROX_PAR_META_DATA
+
+  private[rapids] def compressColumnarBatchWithParquet(
+     gpuCB: ColumnarBatch): List[ParquetCachedBatch] = {
+    // NOTE: this doesn't take nulls into account so we could be over-estimating the acutal size
```
Comment (jlowe) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:354]:
```suggestion
    // NOTE: this doesn't take nulls into account so we could be over-estimating the actual size
```

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 4]
```diff
@@ -330,19 +340,58 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
...
+    // but that's still better than under-estimating
+    val estimatedRowSize = scala.Range(0, gpuCB.numCols()).map { index =>
+      gpuCB.column(index).dataType().defaultSize
+    }.sum
+    val rowsInBatch = (BYTES_ALLOWED_PER_BATCH / estimatedRowSize).toInt
+    val splitIndices = scala.Range(rowsInBatch, gpuCB.numRows(), rowsInBatch)
+    val cudfTables = if (splitIndices.nonEmpty) {
+      val splutVectors = scala.Range(0, gpuCB.numCols()).map { index =>
```
Comment (jlowe) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:362]: Was this supposed to be named `splitVectors`?

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 5]
```diff
@@ -330,19 +340,58 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
...
+  private[rapids] def compressColumnarBatchWithParquet(
+     gpuCB: ColumnarBatch): List[ParquetCachedBatch] = {
+    // NOTE: this doesn't take nulls into account so we could be over-estimating the acutal size
+    // but that's still better than under-estimating
+    val estimatedRowSize = scala.Range(0, gpuCB.numCols()).map { index =>
+      gpuCB.column(index).dataType().defaultSize
+    }.sum
+    val rowsInBatch = (BYTES_ALLOWED_PER_BATCH / estimatedRowSize).toInt
+    val splitIndices = scala.Range(rowsInBatch, gpuCB.numRows(), rowsInBatch)
+    val cudfTables = if (splitIndices.nonEmpty) {
```
Comment (jlowe) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:361]: This seems to be leaked if there's an exception.

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 6]
```diff
@@ -330,19 +340,58 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
...
+      try {
+        // Splitting the table
+        // e.g. T0 = {col1, col2,...,coln} => split columns into 'm' cols =>
+        // T00= {splitCol1(0), splitCol2(0),...,splitColn(0)}
+        // T01= {splitCol1(1), splitCol2(1),...,splitColn(1)}
+        // ...
+        // T0m= {splitCol1(m), splitCol2(m),...,splitColn(m)}
+        for (i <- splutVectors(0).indices) yield
+          new Table({
+            for (j <- splutVectors.indices) yield splutVectors(j)(i)
```
Comment (jlowe) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:374]: This has the drawback of creating a bunch of leakable resources up front. I believe this will be easier to maintain and read if we construct the one table we need to perform the write then close that table rather than creating them all up-front first.  The code that writes a single table can be refactored into a separate method so it's easy to call from this split case and the non-split case rather than needing to fall-through to common-code.

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 7]
```diff
@@ -330,19 +340,58 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
...
+  val _2GB: Long = 2L * 1024 * 1024 * 1024
+  val APPROX_PAR_META_DATA: Int = 10 * 1024 * 1024 // we are estimating 10MB
+  val BYTES_ALLOWED_PER_BATCH: Long = _2GB - APPROX_PAR_META_DATA
+
+  private[rapids] def compressColumnarBatchWithParquet(
+     gpuCB: ColumnarBatch): List[ParquetCachedBatch] = {
+    // NOTE: this doesn't take nulls into account so we could be over-estimating the acutal size
+    // but that's still better than under-estimating
+    val estimatedRowSize = scala.Range(0, gpuCB.numCols()).map { index =>
+      gpuCB.column(index).dataType().defaultSize
```
Comment (jlowe) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:357]: Since these are GPU vectors, wouldn't it be more accurate to estimate the row size based on the actual device memory size and dividing by the number of rows?  `ColumnVector.getDeviceMemorySize()`

File: tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala [hunk 8]
```diff
@@ -118,4 +133,111 @@ class ParquetWriterSuite extends SparkQueryCompareTestSuite {
...
+      withResource(cudfCols) { _=>
+        val cb = new ColumnarBatch(gpuCols, ROWS)
+        whenSplitCalled(cb)
+        try {
+          val method = compressWithParquetMethod.getOrElse {
+            val classOfSerializer = Class.forName(
+              "com.nvidia.spark.rapids.shims.spark310.ParquetCachedBatchSerializer")
+            parquetSerializerInstance = Some(classOfSerializer.newInstance())
+            val compressWithParquet =
+              classOfSerializer.getMethod("compressColumnarBatchWithParquet",
```
Comment (jlowe) [inline tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala:231]: Why is this using reflection to call code we wrote?

**Review by jlowe — CHANGES_REQUESTED** [review]

File: tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala (hunk 8 above)
Comment (razajafri) [reply tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala:231]: Because the code won't compile against Spark 3.0 if we directly reference the `ParquetCachedBatchSerializer`. 

This is why we need a test framework with which we can ignore tests at compile time.

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala (hunk 7 above)
Comment (razajafri) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:357]: Will be addressed in #1285

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala (hunk 6 above)
Comment (razajafri) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:374]: Fair point. This will be addressed in #1286?

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala (hunk 5 above)
Comment (razajafri) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:361]: This will be addressed in #1287

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala (hunk 4 above)
Comment (razajafri) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:362]: This was me being creative, :)  but I can rename this as a follow-on #1288

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala (hunk 2 above)
Comment (razajafri) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:257]: Can we address this as a part of #1289?

**Comment by razajafri** [comment]
I have created follow-ons to address your comments @jlowe

File: tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala (hunk 8 above)
Comment (jlowe) [reply tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala:231]: ah right the tests problem.  that too will be a followup.

**Review by jlowe — DISMISSED** [review]

**Comment by jlowe** [comment]
CI will probably fail without merging up to branch-0.3 which contains the fix for #1275

**Comment by jlowe** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by GaryShen2008** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
Blossom is timing out. Kicking it off again

**Comment by razajafri** [comment]
Build

**Comment by razajafri** [comment]
Build

File: pom.xml [hunk 9]
```diff
@@ -319,7 +319,7 @@
             <dependency>
               <groupId>org.mockito</groupId>
               <artifactId>mockito-core</artifactId>
-              <version>2.28.2</version>
+              <version>3.6.0</version>
```
Comment (revans2) [inline pom.xml:322]: nit: Is this change still needed?

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 10]
```diff
@@ -330,19 +340,58 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
...
-      withResource(Table.writeParquetChunked(ParquetWriterOptions.DEFAULT, buffer)) { writer =>
-        writer.write(table)
+  val _2GB: Long = 2L * 1024 * 1024 * 1024
+  val APPROX_PAR_META_DATA: Int = 10 * 1024 * 1024 // we are estimating 10MB
+  val BYTES_ALLOWED_PER_BATCH: Long = _2GB - APPROX_PAR_META_DATA
+
+  private[rapids] def compressColumnarBatchWithParquet(
+     gpuCB: ColumnarBatch): List[ParquetCachedBatch] = {
+    // NOTE: this doesn't take nulls into account so we could be over-estimating the actual size
+    // but that's still better than under-estimating
```
Comment (revans2) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:355]: But this could underestimate the size.  It makes some assumptions like strings are 10 bytes long, and that arrays hold a single value in them.

Why are we not getting the actual size of the batch and using it guess at the number of rows per batch.  It would still not be perfect, but better than this.

```
estimatedRowSize = GpuColumnVector.getTotalDeviceMemoryUsed(gpuCB) / gpuCB.getNumRows()
```

Might need to check for empty batches, but hopefully we don't have any at this point.

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 11]
```diff
@@ -330,19 +340,58 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
...
+      } finally {
+        for (i <- splutVectors(0).indices)
+            for (j <- splutVectors.indices) splutVectors(j)(i).close()
+      }
+    } else {
+      Seq(GpuColumnVector.from(gpuCB))
+    }
+
+    val buffers = new ListBuffer[ParquetCachedBatch]
+    cudfTables.foreach { table =>
```
Comment (revans2) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:385]: This will leak a number of things in the case of an exception.  I'm not sure it is worth the added complexity right now to fix it, but it would be good to at least put in a comment and file a follow on issue to look into it.

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala [hunk 12]
```diff
@@ -1162,10 +1248,8 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
       val columnarBatchRdd = input.mapPartitions(iter => {
         new RowToColumnarIterator(iter, structSchema, RequireSingleBatch, converters)
       })
-      columnarBatchRdd.map(cb => {
-        withResource(cb) { columnarBatch =>
-          val cachedBatch = compressColumnarBatchWithParquet(columnarBatch)
-          cachedBatch
+      columnarBatchRdd.flatMap(cb => {
+        withResource(cb) { cb => compressColumnarBatchWithParquet(cb)
```
Comment (revans2) [inline shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:1252]: nit: It is probably cleaner to do it like the following.  At a minimum if the closure can fit on a single line it is ideal to use `()` instead of `{}` and to put it on a single line.

```
withResource(cb)(compressColumnarBatchWithParquet)
```

File: tests/pom.xml [hunk 13]
```diff
@@ -104,7 +104,8 @@
         </dependency>
         <dependency>
             <groupId>org.mockito</groupId>
-            <artifactId>mockito-core</artifactId>
+            <artifactId>mockito-inline</artifactId>
+            <version>3.6.0</version>
```
Comment (revans2) [inline tests/pom.xml:108]: If we need a version in multiple places we should have the base pom set a property with that version so we can be consistent.  Also we should have the base POM do the setting of the versions in the dependencyManagement section.

File: tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala [hunk 14]
```diff
@@ -118,4 +125,112 @@ class ParquetWriterSuite extends SparkQueryCompareTestSuite {
...
+          // noop
+        }
+      })
+
+    withResource(cudfCols) { _ =>
+      val cb = new ColumnarBatch(gpuCols, ROWS)
+      whenSplitCalled(cb)
+      try {
+        val method = compressWithParquetMethod.getOrElse {
+          val classOfSerializer = Class.forName(
```
Comment (revans2) [inline tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala:221]: I would prefer to have us file a follow on issue to have tests that compile/run depending on the version of the dependency instead of having us do reflection.

File: pom.xml (hunk 9 above)
Comment (razajafri) [reply pom.xml:322]: yes, for mocking static methods in `Table'

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala (hunk 10 above)
Comment (razajafri) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:355]: Thanks Bobby, an issue has already been filed for this. Please checkout #1285

File: shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala (hunk 11 above)
Comment (razajafri) [reply shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala:385]: #1287 is filed

File: pom.xml (hunk 9 above)
Comment (revans2) [reply pom.xml:322]: I really don't feel good about mocking static methods, but OK.

File: tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala (hunk 14 above)
Comment (razajafri) [reply tests/src/test/scala/com/nvidia/spark/rapids/ParquetWriterSuite.scala:221]: Can we let this one stay? I would rather have some parts tested, I already have filed a follow-on for the test framework
#1290

**Review by revans2 — APPROVED** [review]

**Comment by razajafri** [comment]
build

**Review by revans2 — APPROVED** [review]
